src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -30,53 +30,63 @@ internal class Http2OutputProducer : IHttpOutputProducer, IHttpOutputAborter
         private readonly MemoryPool<byte> _memoryPool;
         private readonly Http2Stream _stream;
         private readonly object _dataWriterLock = new object();
-        private readonly PipeWriter _pipeWriter;
+        private readonly Pipe _pipe;
+        private readonly ConcurrentPipeWriter _pipeWriter;
         private readonly PipeReader _pipeReader;
-        private readonly ValueTask<FlushResult> _dataWriteProcessingTask;
+        private ValueTask<FlushResult> _dataWriteProcessingTask;
         private bool _startedWritingDataFrames;
-        private bool _completed;
+        private bool _streamCompleted;
         private bool _suffixSent;
         private bool _streamEnded;
-        private bool _disposed;
+        private bool _writerComplete;
 
         private IMemoryOwner<byte> _fakeMemoryOwner;
 
-        public Http2OutputProducer(
-            int streamId,
-            Http2FrameWriter frameWriter,
-            StreamOutputFlowControl flowControl,
-            MemoryPool<byte> pool,
-            Http2Stream stream,
-            IKestrelTrace log)
+        public Http2OutputProducer(Http2Stream stream, Http2StreamContext context, StreamOutputFlowControl flowControl)
         {
-            _streamId = streamId;
-            _frameWriter = frameWriter;
-            _flowControl = flowControl;
-            _memoryPool = pool;
             _stream = stream;
-            _log = log;
+            _frameWriter = context.FrameWriter;
+            _flowControl = flowControl;
+            _memoryPool = context.MemoryPool;
+            _log = context.ServiceContext.Log;
 
-            var pipe = CreateDataPipe(pool);
+            _pipe = CreateDataPipe(_memoryPool);
 
-            _pipeWriter = new ConcurrentPipeWriter(pipe.Writer, pool, _dataWriterLock);
-            _pipeReader = pipe.Reader;
+            _pipeWriter = new ConcurrentPipeWriter(_pipe.Writer, _memoryPool, _dataWriterLock);
```

Comment:
I'm actually surprised that ConcurrentPipeWriter doesn't show up in your profile. I guess there aren't many scenarios where it would allocate.

Reply:
For comparison, how much did it allocate before?

Reply:
You can see it here:

![image](https://user-images.githubusercontent.com/303201/75140250-61076c80-5753-11ea-8f28-7707290163b6.png)